README.md
# Tuple dart

A library providing a tuple data structure for Dart. 

* No dependency
* Easy usage
* Support `tuple1` - `tuple8`
* ideal for use with `compute`
* Used by the [itertools](https://pub.dev/packages/itertools) package

## Usage

```dart
import 'package:tuple_dart/tuple.dart';
```
or 
```dart
import 'package:tuple_dart/tuple_dart.dart';
```

## Example

```dart
import 'package:tuple_dart/tuple.dart';

void print2(Object? first, Object? second) {
  print('$first $second');
}

void main() {
  final tuple = Tuple2('first', 2); // type Tuple2<String, int>
  tuple.apply(print2); // output: "first 2";
  final tuple2 = tuple.withItem2('second'); // type Tuple2<String, String>
  tuple2.apply(print2); // output: "first second";
  // or equal
  print2.tuplized(tuple); // output: "first 2";
  // also R Function(Tuple<T1...TN>) has detuplized getter
  
  // tuplize transformation
  // R Function(T1...TN) -> R Function(Tuple<T1...TN>);
  // final print2tuplized = print2.tuplized; 
  
  // detuplized transformation
  // R Function(Tuple<T1...TN>) -> R Function(T1...TN);
  // final print2detupled = print2tuplized.detuplized; 

  // compatible with MapEntry
  final tuples = [
    Tuple2('1', 111),
    Tuple2('2', 222),
    Tuple2('3', 333),
    Tuple2('4', 444),
  ];

  print(Map<String, int>.fromEntries(tuples));
}

```

### ideal for use with compute

```dart
import 'package:tuple_dart/tuple.dart';
import 'package:flutter/foundation.dart';

double hardCalculation(double a, double b, double c) {
  double result = 0.0;
  for (int i = 1; i < 1000; i++) {
    result = i * a + b * c / i;
  }
  return result;
}

void main() async {
  final result = await compute(hardCalculation.tuplized, Tuple3(0.1, 2.0, 10.0));
  print(result);
}
```


## Links

[Pub dev][pubdev]

[Documentation][docs]

[Issue tracker][tracker]

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/avdosev/tuple_dart/issues
[pubdev]: https://pub.dev/packages/tuple_dart
[docs]: https://pub.dev/documentation/tuple_dart/latest/

CHANGELOG.md
## 1.4.2
## 1.4.1

* some fixes with pub.dev

## 1.4.0

* add compatibility with MapEntry
* update readme

## 1.3.0

* add tuplized
* add detuplized
* update readme

## 1.2.0

* fix hashCode
* add example

## 1.1.0

* add Tuple8
* add `apply` method

## 1.0.0

- Initial version.


test/tuple_dart_test.dart
import 'package:tuple_dart/tuple.dart';
import 'package:test/test.dart';

void main() {
  
}


lib/src/tuplize.dart
import 'tuple.dart';
import 'func_decl.dart';

/// extension providing tuplize transformation
/// ```
/// R Function(T1...TN) -> R Function(Tuple<T1...TN>);
/// ```
extension Tuplize1<R, T1> on Func1<R, T1> {
  /// get tupled function
  Func1<R, Tuple1<T1>> get tuplized => (Tuple1<T1> tuple) => tuple.apply(this);
}

/// extension providing detuplize transformation
/// ```
/// R Function(Tuple<T1...TN>) -> R Function(T1...TN);
/// ```
extension Detuplize1<R, T1> on Func1<R, Tuple1<T1>> {
  /// get detupled function
  Func1<R, T1> get detuplized => (p0) => this.call(Tuple1(p0));
}

/// extension providing tuplize transformation
/// ```
/// R Function(T1...TN) -> R Function(Tuple<T1...TN>);
/// ```
extension Tuplize2<R, T1, T2> on Func2<R, T1, T2> {
  /// get tupled function
  Func1<R, Tuple2<T1, T2>> get tuplized =>
      (Tuple2<T1, T2> tuple) => tuple.apply(this);
}

/// extension providing detuplize transformation
/// ```
/// R Function(Tuple<T1...TN>) -> R Function(T1...TN);
/// ```
extension Detuplize2<R, T1, T2> on Func1<R, Tuple2<T1, T2>> {
  /// get detupled function
  Func2<R, T1, T2> get detuplized => (p0, p1) => this.call(Tuple2(p0, p1));
}

/// extension providing tuplize transformation
/// ```
/// R Function(T1...TN) -> R Function(Tuple<T1...TN>);
/// ```
extension Tuplize3<R, T1, T2, T3> on Func3<R, T1, T2, T3> {
  /// get tupled function
  Func1<R, Tuple3<T1, T2, T3>> get tuplized =>
      (Tuple3<T1, T2, T3> tuple) => tuple.apply(this);
}

/// extension providing detuplize transformation
/// ```
/// R Function(Tuple<T1...TN>) -> R Function(T1...TN);
/// ```
extension Detuplize3<R, T1, T2, T3> on Func1<R, Tuple3<T1, T2, T3>> {
  /// get detupled function
  Func3<R, T1, T2, T3> get detuplized =>
      (p0, p1, p2) => this.call(Tuple3(p0, p1, p2));
}

/// extension providing tuplize transformation
/// ```
/// R Function(T1...TN) -> R Function(Tuple<T1...TN>);
/// ```
extension Tuplize4<R, T1, T2, T3, T4> on Func4<R, T1, T2, T3, T4> {
  /// get tupled function
  Func1<R, Tuple4<T1, T2, T3, T4>> get tuplized =>
      (Tuple4<T1, T2, T3, T4> tuple) => tuple.apply(this);
}

/// extension providing detuplize transformation
/// ```
/// R Function(Tuple<T1...TN>) -> R Function(T1...TN);
/// ```
extension Detuplize4<R, T1, T2, T3, T4> on Func1<R, Tuple4<T1, T2, T3, T4>> {
  /// get detupled function
  Func4<R, T1, T2, T3, T4> get detuplized =>
      (p0, p1, p2, p3) => this.call(Tuple4(p0, p1, p2, p3));
}

/// extension providing tuplize transformation
/// ```
/// R Function(T1...TN) -> R Function(Tuple<T1...TN>);
/// ```
extension Tuplize5<R, T1, T2, T3, T4, T5> on Func5<R, T1, T2, T3, T4, T5> {
  /// get tupled function
  Func1<R, Tuple5<T1, T2, T3, T4, T5>> get tuplized =>
      (Tuple5<T1, T2, T3, T4, T5> tuple) => tuple.apply(this);
}

/// extension providing detuplize transformation
/// ```
/// R Function(Tuple<T1...TN>) -> R Function(T1...TN);
/// ```
extension Detuplize5<R, T1, T2, T3, T4, T5>
    on Func1<R, Tuple5<T1, T2, T3, T4, T5>> {
  /// get detupled function
  Func5<R, T1, T2, T3, T4, T5> get detuplized =>
      (p0, p1, p2, p3, p4) => this.call(Tuple5(p0, p1, p2, p3, p4));
}

/// extension providing tuplize transformation
/// ```
/// R Function(T1...TN) -> R Function(Tuple<T1...TN>);
/// ```
extension Tuplize6<R, T1, T2, T3, T4, T5, T6>
    on Func6<R, T1, T2, T3, T4, T5, T6> {
  /// get tupled function
  Func1<R, Tuple6<T1, T2, T3, T4, T5, T6>> get tuplized =>
      (Tuple6<T1, T2, T3, T4, T5, T6> tuple) => tuple.apply(this);
}

/// extension providing detuplize transformation
/// ```
/// R Function(Tuple<T1...TN>) -> R Function(T1...TN);
/// ```
extension Detuplize6<R, T1, T2, T3, T4, T5, T6>
    on Func1<R, Tuple6<T1, T2, T3, T4, T5, T6>> {
  /// get detupled function
  Func6<R, T1, T2, T3, T4, T5, T6> get detuplized =>
      (p0, p1, p2, p3, p4, p5) => this.call(Tuple6(p0, p1, p2, p3, p4, p5));
}

/// extension providing tuplize transformation
/// ```
/// R Function(T1...TN) -> R Function(Tuple<T1...TN>);
/// ```
extension Tuplize7<R, T1, T2, T3, T4, T5, T6, T7>
    on Func7<R, T1, T2, T3, T4, T5, T6, T7> {
  /// get tupled function
  Func1<R, Tuple7<T1, T2, T3, T4, T5, T6, T7>> get tuplized =>
      (Tuple7<T1, T2, T3, T4, T5, T6, T7> tuple) => tuple.apply(this);
}

/// extension providing detuplize transformation
/// ```
/// R Function(Tuple<T1...TN>) -> R Function(T1...TN);
/// ```
extension Detuplize7<R, T1, T2, T3, T4, T5, T6, T7>
    on Func1<R, Tuple7<T1, T2, T3, T4, T5, T6, T7>> {
  /// get detupled function
  Func7<R, T1, T2, T3, T4, T5, T6, T7> get detuplized =>
      (p0, p1, p2, p3, p4, p5, p6) =>
          this.call(Tuple7(p0, p1, p2, p3, p4, p5, p6));
}

/// extension providing tuplize transformation
/// ```
/// R Function(T1...TN) -> R Function(Tuple<T1...TN>);
/// ```
extension Tuplize8<R, T1, T2, T3, T4, T5, T6, T7, T8>
    on Func8<R, T1, T2, T3, T4, T5, T6, T7, T8> {
  /// get tupled function
  Func1<R, Tuple8<T1, T2, T3, T4, T5, T6, T7, T8>> get tuplized =>
      (Tuple8<T1, T2, T3, T4, T5, T6, T7, T8> tuple) => tuple.apply(this);
}

/// extension providing detuplize transformation
/// ```
/// R Function(Tuple<T1...TN>) -> R Function(T1...TN);
/// ```
extension Detuplize8<R, T1, T2, T3, T4, T5, T6, T7, T8>
    on Func1<R, Tuple8<T1, T2, T3, T4, T5, T6, T7, T8>> {
  /// get detupled function
  Func8<R, T1, T2, T3, T4, T5, T6, T7, T8> get detuplized =>
      (p0, p1, p2, p3, p4, p5, p6, p7) =>
          this.call(Tuple8(p0, p1, p2, p3, p4, p5, p6, p7));
}


scripts/write_tuple.dart
import 'dart:io';

String makeTupleCode(int length) {
  final types = Iterable.generate(length, (int i) => 'T${i + 1}').toList();
  final fields = Iterable.generate(length, (int i) => 'item${i + 1}').toList();
  final classFields =
      Iterable.generate(length, (int i) => '  final T${i + 1} item${i + 1};\n')
          .join();

  final withMethods = Iterable.generate(length, (int i) {
    final t = [...types.take(i), 'E', ...types.skip(i + 1)];
    final f = [...fields.take(i), 'newItem', ...fields.skip(i + 1)];
    return """
  /// Returns a tuple with the item${i + 1} set to the specified value.
  Tuple$length<${t.join(', ')}> withItem${i + 1}<E>(E newItem) => Tuple$length(${f.join(', ')});
""";
  });

  return """/// Represents a $length-tuple
class Tuple$length<${types.join(', ')}> {
$classFields
  const Tuple$length(${fields.map((e) => 'this.$e').join(', ')});

${withMethods.join()}
  /// apply a function `f` over tuple fields
  R apply<R>(R Function(${types.join(', ')}) f) {
    return f(${fields.join(', ')});
  }

  Iterable get iterable sync* {
${fields.map((e) => '    yield $e;\n').join()}  }

  List toList({bool growable = false}) => List.from(iterable, growable: growable);

  @override
  String toString() => '(${fields.map((e) => '\$$e').join(', ')})';

  @override
  bool operator ==(Object other) =>
      other is Tuple$length &&
${fields.map((e) => '      other.$e == $e').join(' &&\n')};

  @override
  int get hashCode => Object.hash(${fields.join(', ')});
}
""";
}

void main(List<String> args) async {
  final f = File('lib/src/tuple.dart');
  final file = f.openWrite();
  for (var i = 1; i <= 8; i++) {
    file.write(makeTupleCode(i));
    if (i != 8) file.write('\n');
  }
  await file.flush();
  await file.close();
}


scripts/write_tuplize.dart
import 'dart:io';

Future<void> fncDcl() async {
  final f = File('lib/src/func_decl.dart');
  final file = f.openWrite();
  for (var i = 0; i <= 8; i++) {
    file.write(makeFuncDeclCode(i));
  }
  await file.flush();
  await file.close();
}

Future<void> tplz() async {
  final f = File('lib/src/tuplize.dart');
  final file = f.openWrite();
  file.writeln("import 'tuple.dart';");
  file.writeln("import 'func_decl.dart';");
  file.writeln();
  for (var i = 1; i <= 8; i++) {
    file.writeln(makeTupleCode(i));
  }
  await file.flush();
  await file.close();
}

void main(List<String> args) async {
  await Future.wait([fncDcl(), tplz()]);
}

String makeFuncDeclCode(int length) {
  return """
typedef Func$length<R${List.generate(length, (index) => ', T${index + 1}').join()}> = R Function(${List.generate(length, (index) => 'T${index + 1}').join(', ')});
""";
}

String makeTupleCode(int length) {
  final types = List.generate(length, (index) => 'T${index + 1}').join(', ');
  final sep_types = List.generate(length, (index) => ', T${index + 1}').join();
  final types_arg =
      List.generate(length, (index) => 'T${index + 1} p$index').join(', ');
  final args = List.generate(length, (index) => 'p$index').join(', ');
  return """
/// extension providing tuplize transformation
/// ```
/// R Function(T1...TN) -> R Function(Tuple<T1...TN>);
/// ```
extension Tuplize$length<R$sep_types> on Func$length<R$sep_types> {
  /// get tupled function
  Func1<R, Tuple$length<$types>> get tuplized => (Tuple$length<$types> tuple) => tuple.apply(this); 
}

/// extension providing detuplize transformation
/// ```
/// R Function(Tuple<T1...TN>) -> R Function(T1...TN);
/// ```
extension Detuplize$length<R$sep_types> on Func1<R, Tuple$length<$types>> {
  /// get detupled function
  Func$length<R$sep_types> get detuplized => ($args) => this.call(Tuple$length($args)); 
}
""";
}


example/example.dart
import 'package:tuple_dart/tuple.dart';
// or import 'package:tuple_dart/tuple_dart.dart';

void print2(Object? first, Object? second) {
  print('$first $second');
}

void main() {
  final tuple = Tuple2('first', 2); // type Tuple2<String, int>
  tuple.apply(print2); // output: "first 2";
  final tuple2 = tuple.withItem2('second'); // type Tuple2<String, String>
  tuple2.apply(print2); // output: "first second";
  // or equal
  print2.tuplized(tuple); // output: "first 2";
  // also R Function(Tuple<T1...TN>) has detuplized getter

  // tuplize transformation
  // R Function(T1...TN) -> R Function(Tuple<T1...TN>);
  // final print2tuplized = print2.tuplized;

  // detuplized transformation
  // R Function(Tuple<T1...TN>) -> R Function(T1...TN);
  // final print2detupled = print2tuplized.detuplized;

  // compatible with MapEntry
  final tuples = [
    Tuple2('1', 111),
    Tuple2('2', 222),
    Tuple2('3', 333),
    Tuple2('4', 444),
  ];

  print(Map<String, int>.fromEntries(tuples));
}
